### ISSUE TITLE
Return a custom error from an openapi endpoint

### ISSUE THREAD

**Issue by ufoscout** [poem-web/poem#244]
I have some endpoints implemented in `poem` that I would like to migrate to `poem-openapi`.
All the endpoints return a `Result` with the same custom error implementing `ResponseError`.
For example:
```rust 
use thiserror::Error;

#[derive(Error, Debug)]
pub enum CustomError {
    #[error("BadRequest")]
    BadRequest,
    #[error("Internal")]
    Internal,
}

impl ResponseError for CustomError {
    fn status(&self) -> StatusCode {
        match self {
            CustomError::BadRequest => StatusCode::FORBIDDEN,
            CustomError::Internal =>  StatusCode::INTERNAL_SERVER_ERROR,
        }
    }
}

#[handler]
async fn get() -> Result<(), CustomError> {
    Ok(())
}
```

This works fine in `poem`; nevertheless, this does not compile with `poem-openapi`:
```rust
struct Api;

//Error: the trait bound `Result<(), CustomError>: ApiResponse` is not satisfied
#[OpenApi] 
impl Api {

    #[oai(path = "/ok", method = "get")]
    async fn web_nok(&self) -> Result<(), CustomError> {
        Ok(())
    }

}
```

I can make it compile with this workaround:
```rust
struct Api;

#[OpenApi] 
impl Api {

    #[oai(path = "/ok", method = "get")]
    async fn web_nok(&self) -> poem::Result<()> {
        Err(CustomError::BadRequest)?
    }

}
```
But, in this case, the response type and code are not mapped in the swagger UI.
Is there a way to share the same custom error between the two different types of endpoints and expose the expected types and code in the swagger UI?

**Comment by Christoph-AK** [comment]
See https://github.com/poem-web/poem/issues/230#issuecomment-1065112886

I have a function in my error struct to return the error as Json<Whatever>.

**Comment by ufoscout** [comment]
@Christoph-AK thanks, but I wonder if there's a more straightforward way of doing it.

**Comment by ufoscout** [comment]
@sunli829 Would it be possible to allow an OpenApi endpoint to return whatever result with an Error that implements `ApiResponse`?
So it would be possible to write:
```rust
use poem::web::Json;
use poem_openapi::{OpenApi, ApiResponse};
use serde::Serialize;
use thiserror::Error;


#[derive(Error, Debug)]
pub enum CustomError {
    #[error("BadRequest")]
    BadRequest,
    #[error("Internal")]
    Internal,
}

#[derive(ApiResponse)]
pub enum ApiErrorResponse {
    #[oai(status = 400)]
    BadRequest,
    #[oai(status = 500)]
    InternalServerError,
}

impl From<CustomError> for ApiErrorResponse {
    fn from(err: CustomError) -> Self {
        match err {
            CustomError::BadRequest => ApiErrorResponse::BadRequest,
            CustomError::Internal =>  ApiErrorResponse::InternalServerError,
        }
    }
}

#[derive(Serialize)]
struct MyJson {}

struct Api;

#[OpenApi] 
impl Api {

    // Here you can return a result
    // If the result is Ok(), then the response status code is 200 and the type Json<MyJson>
    // If the result is Err(), the the response status code and type are defined by ApiErrorResponse
    #[oai(path = "/json", method = "get")]
    async fn get_json(&self) -> Result<Json<MyJson>, ApiErrorResponse> {
        // You can now use the ? operator
        Err(CustomError::BadRequest)?;
        Ok(Json(MyJson{}))
    }

}

```

I think this would be very ergonomic to use.

**Comment by sunli829** [comment]
I think it's a good idea. I'll try it right now!

**Comment by sunli829** [comment]
Released in `v1.3.17`

Here is the example:

https://github.com/poem-web/poem/blob/aac66274eed5c46e77d1c8bd469c9e5e8211aba1/poem-openapi/tests/response.rs#L406

**Comment by ufoscout** [comment]
@sunli829 
Thanks for this! I am testing `1.3.17` and, although it compiles, it does not generate the expected response codes and types in the swagger UI:
![Screenshot_20220323_141304](https://user-images.githubusercontent.com/960885/159707991-e5ccac34-17ec-44ce-88c2-367494be7d9c.png)

In fact, I expected to find also the 400 and 500 status codes and related data schemas.

Here is the full code of my attempt:
```rust
use poem::{http::StatusCode, error::ResponseError, Server, Route, listener::TcpListener};
use poem_openapi::{OpenApi, OpenApiService, ApiResponse, Object, payload::Json };
use serde::Serialize;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum CustomError {
    #[error("BadRequest")]
    BadRequest,
    #[error("Internal")]
    Internal,
}

impl ResponseError for CustomError {
    fn status(&self) -> StatusCode {
        match self {
            CustomError::BadRequest => StatusCode::BAD_REQUEST,
            CustomError::Internal =>  StatusCode::INTERNAL_SERVER_ERROR,
        }
    }
}

#[derive(Serialize, Object)]
pub struct MyJson {
    a: String
}

#[derive(Serialize, Object, Debug)]
pub struct InternalServerErrorPayload {
    message: String
}

#[derive(ApiResponse)]
pub enum ApiJsonResponse {
    #[oai(status = 200)]
    Ok(Json<MyJson>),
}

#[derive(ApiResponse, Debug)]
pub enum ApiErrorResponse {
    #[oai(status = 400)]
    BadRequest,
    #[oai(status = 500)]
    InternalServerError(Json<InternalServerErrorPayload>),
}

impl From<CustomError> for ApiErrorResponse {
    fn from(err: CustomError) -> Self {
        match err {
            CustomError::BadRequest => ApiErrorResponse::BadRequest,
            CustomError::Internal =>  ApiErrorResponse::InternalServerError(Json(InternalServerErrorPayload{ message: "something bad".to_owned()})),
        }
    }
}

struct Api;

#[OpenApi] 
impl Api {
    #[oai(path = "/json", method = "get")]
    async fn get_json(&self) -> Result<ApiJsonResponse, ApiErrorResponse> {
        // You can now use the ? operator
        Err(CustomError::BadRequest)?;
        Ok(ApiJsonResponse::Ok(Json(MyJson{a: "hello".to_owned()})))
    }
}

#[tokio::main]
async fn main() -> Result<(), std::io::Error> {
    if std::env::var_os("RUST_LOG").is_none() {
        std::env::set_var("RUST_LOG", "poem=debug");
    }

    let api_service =
        OpenApiService::new(Api, "Hello World", "1.0").server("http://localhost:3000/api");
    let ui = api_service.swagger_ui();

    Server::new(TcpListener::bind("localhost:3000"))
        .run(Route::new().nest("/api", api_service).nest("/", ui))
        .await
}
```

**Comment by sunli829** [comment]
Since Rust does not yet support specialization, this requires some macro changes. I'll add it tomorrow.

**Comment by sunli829** [comment]
Released in `v1.3.18` 😁

**Comment by ufoscout** [comment]
@sunli829 Amazing!! Thank you!
